Dewikification - Implement redirections app

ks129 opened the pull request:
This PR adds a new `redirect` app whose responsibility is ensuring that after Dewikification links still work. I had to create a custom redirect view to add functionality for our use case.

There is possible to define 2 types of redirects: exact redirects and prefix redirects.

There are survival tests for every redirect route that ensures:
1. There is only 1 redirect.
2. Redirects to the given URL.
3. Final destination response code is 200.

Add resources and events app redirects, ~~content apps redirects still have to be added (except contributing prefix redirect)~~.

ks129 committed:
Create basic redirect app

ks129 committed:
Add redirect app to settings and add resources redirections config

ks129 committed:
Create URLs file for redirection app that dynamically generates routes

ks129 committed:
Include redirect app URLs to home app URLs

ks129 committed:
Add dynamic tests for redirections

ks129 committed:
Add events index redirection

ks129 committed:
Create custom RedirectView to support static arguments

ks129 committed:
Add missing type annotation to get_redirect_url

ks129 committed:
Migrate URLs generation from Django's own RedirectView -> CustomRedirectView

ks129 committed:
Migrate redirections definition to not include testing args

ks129 committed:
Update redirect app tests to take account recent changes

ks129 committed:
Add groups with comments to redirections definition

ks129 committed:
Add code jam pages redirects

ks129 committed:
Migrate redirections from settings.py -> redirects.yaml

ks129 committed:
Implement hacky solution to allow prefix redirects

ks129 committed:
Add contributing pages prefix redirection and update tests

ks129 committed:
Add note about prefix redirects

ks129 committed:
Add code jams helper pages redirects

ks129 committed:
Add code jam 6 and 7 rules pages redirects

ks129 committed:
Add HTTP 200 requirement to redirect survival tests

ks129 committed:
Add game jam pages redirects

kosayoda committed:
Add content app redirects.

lemonsaurus reviewed (commented):
Looks like an overall reasonable and well-coded approach. Let's just make sure we've got everything we talked about in #dev-core and I think this is good to merge.

kosayoda committed:
Route navbar links to their new locations.

kosayoda committed:
Use temporary redirects for now.

kosayoda committed:
Add disclaimer for the redirects app.

kosayoda committed:
Add privacy policy information and link.

This should be redirected to the notion page using our cloudflare
worker.

kosayoda committed:
Make tests expect 302.

It's what we're using now for dewikification redirects, this has to be
changed back to 301 when we decide to make redirects permanent.

kosayoda reviewed (approved):
Would be nice to have one more set of eyes and testing over this.
